articles/user-guide/host-programs.md

```diff
@@ -0,0 +1,455 @@
+Indeed, this process is at the core of the [quantum random number generator](xref:microsoft.quantum.quickstarts.qrng) quickstart.
+
+In Q#, this would be performed by the following code:
+
+```qsharp
+        using (q = Qubit()) {    // allocates qubit for use (automatically in |0>)
+            H(q);                // puts qubit in superposition of |0> and |1>
+            return MResetZ(q);   // measures qubit, returns result (and resets it to |0> before deallocation)
+        }
+```
+
+However, this code alone can't be executed by Q#.
+For that, it needs to make up the body of an [operation](xref:microsoft.quantum.guide.basics#q-operations-and-functions), which is then executed when called---either directly or by another operation. 
+Hence it takes the form
+```qsharp
+    operation MeasureSuperposition() : Result {
+        using (q = Qubit()) {
+            H(q);
+            return MResetZ(q);
+        }
+    }
+```
+and we now have defined an operation, `MeasureSuperposition`, which takes no arguments and returns a value of type [Result](xref:microsoft.quantum.guide.types).
+More details on defining callables can be found at [Operations and functions](xref:microsoft.quantum.guide.operationsfunctions).
+For the sake of brevity we refer primarily to Q# *operations* on this page, however all of the concepts discussed work in the same manner for Q# *functions*, as well as any user-defined Q# types. 
+
+### Operation defined in a Q# file
+
+The operation is precisely what's called and run by Q#.
+However, it requires a few more additions to comprise a full `*.qs` Q# file.
+
+All Q# types and callables are defined within *namespaces*, which the compiler can then open individually 
+
+Firstly, the [`H`](xref:microsoft.quantum.intrinsic.h) and [`MResetZ`](xref:microsoft.quantum.measurement.mresetz) actually belong to the [Q# Standard Libraries](xref:microsoft.quantum.qsharplibintro).
+To be available for use, their respective *namespaces* need to be opened.
+All Q# callables and types (both those you define and those intrinsic to the language) are defined within namespaces.
+To access them, the compiler needs to be made aware of which namespaces to open.
+
+So, the full Q# file containing our operation consists of defining our own namespace, opening the namespaces for the operations our operation uses, and then our operation:
+
+```qsharp
+namespace NamespaceName {
+    open Microsoft.Quantum.Intrinsic;     // for the H operation
+    open Microsoft.Quantum.Measurement;   // for MResetZ
+
+    operation MeasureSuperposition() : Result {
+        using (q = Qubit()) { 
+            H(q);
+            return MResetZ(q);
+        }
+    }
+}
+```
+
+### Execution of a Q# program
+
+Now the general execution model of a Q# program becomes clear.
+
+<br/>
+<img src="./images/general_execution_model.png" alt="Q# program execution diagram" width="400">
```

Comment:
I've generally used a similar figure in talks and presentations: 
![image](https://user-images.githubusercontent.com/31516/82469436-dba50700-9a78-11ea-8206-4aa0ad6b9da1.png)
I think the main thing I'd raise is that namespaces don't directly enter program flow, as they're only used to give names to each operation and function.

Reply:
@geduardo What are you referring to with the "sets" relationship? And _if_ we were to dive into the compiler more, I would definitely want some dev input on the details. I agree that it would need a more complete description, but I'm unsure if that would over-clutter the page even more; especially considering that this page is primarily to detail how to use Q# in the different command-line applications/host programs. When writing the execution model stuff, I tried to err on the "need to know" side, trying to restrict it to the basic ideas that let the user see what the difference is between the environments. I'm open to going either way.

Reply:
I'm still not quite sure what the arrow between "Q# operation" and "Q# namespace" indicates here. Namespaces aren't callable and have no effect on execution flow other than to assign names (as distinct from Python modules, which do have an impact on execution flow, since `import` statements can have arbitrary side effects).

![image](https://user-images.githubusercontent.com/31516/84529824-f85ae600-ac96-11ea-8e61-40df66c4ae9a.png)

Reply:
@cgranade the arrow from the called Q# operation to inside the same namespace is meant to show that the operation has automatic access to all the types and callables defined within the same namespace (as described in the text). I think it aligns with the same functionality as the arrow to "Q# Libraries", indicating access to types and callables defined therein---i.e. not implying that the library itself is callable. Let me know if you think more clarification is needed, and if so, in what manner.